Inline harmonic analysis

c2xu opened the pull request:
The module MOM_HA computes the constant coefficients of sine/cosine functions of each tidal constituent, for the sea surface height and barotropic velocity fields.

herrwang0 commented:
This is a nice addition! 

Comment from a quick glance: Instead of barotropic solver, I wonder if it would be better to apply the calculation to the instantaneous velocity `[uv]` and `eta` at the end of baroclinic corrector step? As it stands, `HA_accum` adds up fields from both predictor and corrector steps.

Hallberg-NOAA reviewed (changes requested):
Thank you for this contribution, @c2xu. This will provide a valuable new capability, and it is off to a great start, but there are a few things that you might want to consider, and a few points where this is tripping over some of our automated code style testing, as documented in the MOM6 style guide (https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide).

  First, this is getting a red-X in the testing because it has trailing blanks and because some of the real variables are not described in documentation. Trailing blanks are easy enough to find by running `grep -n ' $' *.F90` in the directory with the files you are scanning.  We also require that all real arguments and elements in defined types have a short written description of their purpose, including their units and sign conventions as appropriate, in dOxygen formatted comments, and we strongly encourage that all other real variables are also documented in comments (this may become a requirement once the existing code is compliant).  Please add these short descriptions of your variables - others reading the code will appreciate it, and you probably will too if you come back to look at it after having been away from the code for a while!

  As for the code itself, I have a few suggestions:

1. Rather than using real variables for `HA_start_time` and `HA_end_time`, these would be much cleaner and clearer as `time_type` variables.  There are a number of other examples using `time_type` variables in the code that should be a good pattern for you to follow. 
2. As we describe in the style guide, we do not allow the use of module variables.  Please define an appropriate control structure in which to place the variables on lines 38-39 and 58-62 and then pass that around as an argument to the routines that set and use it.  Doing so make it vastly easier to keep track of where data is set and used.
3. We strongly discourage making types transparent, as this PR does with the `tidal_forcing_CS` in MOM_tidal_forcing.F90, unless there is no way to avoid doing so.  If you need information from this control structure, we could add a query function, or perhaps the fact that the new `MOM_HA` module needs to use so much of the information from the tidal forcing module suggests that these probably should be in the same module.
4. Please do not use `MOM_error(WARNING, ...)` messages to note the times of the analysis.  These messages will be issued by every processor (potentially thousands of them), and they really are not warnings of errors that need to be dealt with.  Instead use `MOM_mesg()` or equivalently `MOM_error(NOTE, ...)` for these informational messages, and only issue these notes if you would be doing the harmonic analysis because `HA_end_time > HA_start_time`.
5. This new capability is an unusual case in that the actions are based on times relative to the start of run segments, rather than absolute model times, so the behavior will depend on when the model has undergone a restart.  This code is also doing analysis that can not span restarts without giving different answers.  I can not think of anything else in the MOM6 code that works this way.  Is there a way that these analyses could be triggered based on absolute model times to avoid this behavior (there might not be)?
6. The name `MOM_HA.F90` is not very descriptive.  Perhaps `MOM_harmonic_analysis.F90` would be more self-explanatory?  Also, would it make more sense to put this new module in the `diagnostics` directory rather than the `parameterizations/lateral` directory?

c2xu commented:
> This is a nice addition!
> 
> Comment from a quick glance: Instead of barotropic solver, I wonder if it would be better to apply the calculation to the instantaneous velocity `[uv]` and `eta` at the end of baroclinic corrector step? As it stands, `HA_accum` adds up fields from both predictor and corrector steps.

In the new commit I placed the accumulating calls after the loop for barotropic time stepping. `HA_accum` should adds up fields from corrector steps now.

c2xu commented:
> Thank you for this contribution, @c2xu. This will provide a valuable new capability, and it is off to a great start, but there are a few things that you might want to consider, and a few points where this is tripping over some of our automated code style testing, as documented in the MOM6 style guide (https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide).
> 
> First, this is getting a red-X in the testing because it has trailing blanks and because some of the real variables are not described in documentation. Trailing blanks are easy enough to find by running `grep -n ' $' *.F90` in the directory with the files you are scanning. We also require that all real arguments and elements in defined types have a short written description of their purpose, including their units and sign conventions as appropriate, in dOxygen formatted comments, and we strongly encourage that all other real variables are also documented in comments (this may become a requirement once the existing code is compliant). Please add these short descriptions of your variables - others reading the code will appreciate it, and you probably will too if you come back to look at it after having been away from the code for a while!
> 
> As for the code itself, I have a few suggestions:
> 
> 1. Rather than using real variables for `HA_start_time` and `HA_end_time`, these would be much cleaner and clearer as `time_type` variables.  There are a number of other examples using `time_type` variables in the code that should be a good pattern for you to follow.
> 2. As we describe in the style guide, we do not allow the use of module variables.  Please define an appropriate control structure in which to place the variables on lines 38-39 and 58-62 and then pass that around as an argument to the routines that set and use it.  Doing so make it vastly easier to keep track of where data is set and used.
> 3. We strongly discourage making types transparent, as this PR does with the `tidal_forcing_CS` in MOM_tidal_forcing.F90, unless there is no way to avoid doing so.  If you need information from this control structure, we could add a query function, or perhaps the fact that the new `MOM_HA` module needs to use so much of the information from the tidal forcing module suggests that these probably should be in the same module.
> 4. Please do not use `MOM_error(WARNING, ...)` messages to note the times of the analysis.  These messages will be issued by every processor (potentially thousands of them), and they really are not warnings of errors that need to be dealt with.  Instead use `MOM_mesg()` or equivalently `MOM_error(NOTE, ...)` for these informational messages, and only issue these notes if you would be doing the harmonic analysis because `HA_end_time > HA_start_time`.
> 5. This new capability is an unusual case in that the actions are based on times relative to the start of run segments, rather than absolute model times, so the behavior will depend on when the model has undergone a restart.  This code is also doing analysis that can not span restarts without giving different answers.  I can not think of anything else in the MOM6 code that works this way.  Is there a way that these analyses could be triggered based on absolute model times to avoid this behavior (there might not be)?
> 6. The name `MOM_HA.F90` is not very descriptive.  Perhaps `MOM_harmonic_analysis.F90` would be more self-explanatory?  Also, would it make more sense to put this new module in the `diagnostics` directory rather than the `parameterizations/lateral` directory?

In the new commit I fixed all issues except for No. 5. I also changed the way how HA_start_time and HA_end_time are defined. The latter is now aligned with the end of the run segment so the user only needs to specify the former, which either determines the start time of HA, if positive, or the length of HA, if negative. (This is Ed's original idea.)

It is relatively more complicated to implement the restarting capability, in terms of both coding and how the start and end time of harmonic analysis should be determined. Specifically, I can think of situations where a user wants the harmonic analysis to be performed, let's say, each month, to see the seasonal change of tides. With the current version of the code, he/she can do so by restarting the simulation every month. If the harmonic analysis can span multiple run segments, then it will be more complicated to implement the capability of performing multiple harmonic analyses with a single simulation. Either way, I'm open to discussion on this point.

Hallberg-NOAA commented on src/diagnostics/MOM_harmonic_analysis.F90:
```diff
@@ -0,0 +1,411 @@
+!> The private control structure for storing the HA info of a particular field
+type, private :: HA_type
+  character(len=16) :: key = "none"          !< Name of the field of which harmonic analysis is to be performed
+  real :: old_time = -1.0                    !< The time of the previous accumulating step
```
Please modify the comments describing the units of the real variables on lines 25 and 27-29.  If these are used in various ways, this description might be `[various]`, but if you could give a specific example (e.g., for SSH) those units would also help in making sense of the intended purpose of these variables.

Hallberg-NOAA commented on src/diagnostics/MOM_harmonic_analysis.F90:
```diff
@@ -0,0 +1,411 @@
+end type HA_node
+
+!> The public control structure of the MOM_harmonic_analysis module
+type, public :: HA_CS ; private
```
It might be clearer to name this type `harmonic_analysis_CS` for clarity and to emulate an approximate naming convention used elsewere.

Hallberg-NOAA commented on src/diagnostics/MOM_harmonic_analysis.F90:
```diff
@@ -0,0 +1,411 @@
+  endif
+
+  ! Determine CS%time_start
+  call get_param(param_file, mdl, "HA_start_time", HA_start_time, &
```
Because our param_file parser is case sensitive, we have adopted the (perhaps poorly documented) convention that we use all upper case for parameters, so in this case "HA_START_TIME" would be preferable over "HA_start_time".

Hallberg-NOAA commented on src/diagnostics/MOM_harmonic_analysis.F90:
```diff
@@ -0,0 +1,411 @@
+  type(HA_type), pointer :: ha1
+  type(HA_node), pointer :: tmp
+  real :: now                               !< The relative time compared with the tidal reference [T ~> s]
+  real :: dt                                !< The current time step size of the accumulator
```
Please add a description of the units of dt, probably [T ~> s].

Hallberg-NOAA commented on src/diagnostics/MOM_harmonic_analysis.F90:
```diff
@@ -0,0 +1,411 @@
+subroutine HA_solver(ha1, nc, FtSSHw)
+  type(HA_type), pointer,              intent(in)  :: ha1
+  integer,                             intent(in)  :: nc
+  real, dimension(:,:,:), allocatable, intent(out) :: FtSSHw !< Work array for Cholesky decomposition
```
Please describe the units of FtSSHw, tmp and FtFw.  In this case, these might be arbitrary units (which we sometimes describe with `[A]`, but if they all have the same arbitrary units, that is useful to note, and all would use the same `[A]` descriptor.

Hallberg-NOAA commented on src/diagnostics/MOM_harmonic_analysis.F90:
```diff
@@ -0,0 +1,411 @@
+  type(HA_CS),            intent(in) :: CS         !< Control structure of the MOM_harmonic_analysis module
+
+  ! Local variables
+  real, dimension(:,:,:), allocatable :: FtSSHw    !< An array containing the harmonic constants
```
What are the units of `FtSSHw`?  If they are arbitrary, please indicate this with a phrase like `in arbitrary units [A]`.

Hallberg-NOAA reviewed (changes requested):
Thank you for this revised PR.  I think that this is getting close to being ready, but there are a number of specific changes (noted in separate comments) that I think should be considered to help with the documentation of parts of this new contribution and its variables.

herrwang0 commented:
> > This is a nice addition!
> > Comment from a quick glance: Instead of barotropic solver, I wonder if it would be better to apply the calculation to the instantaneous velocity `[uv]` and `eta` at the end of baroclinic corrector step? As it stands, `HA_accum` adds up fields from both predictor and corrector steps.
> 
> In the new commit I placed the accumulating calls after the loop for barotropic time stepping. `HA_accum` should adds up fields from corrector steps now.

In the current RK2 scheme, `btstep`() is called twice in both predictor and corrector steps. So, if I understand correctly, as it is coded now, `FtSSH` would include two sets of `[uv]bt` and `eta_bt` for each timestamp, one from the predictor step and one from the corrector step. (similarly, `FtF` would add the trigonometric terms of the same timestamps twice). 

I guess I would move the `HA_acccum`() calls to the end of `step_MOM_dyn_split_RK2`() (and the two unsplit dyn-cores if you like). Barotropic velocity will need to be calculated (`btstep`() does not output velocity, and I wonder if this was partly the reason you placed `HA_acccum`() in barotropic solver), but it is still cheaper than the calculation by including both predictor and corrector velocities. (and more accurate). 

Another related note, it looks like that if harmonic analysis is needed more than one field, `FtF` will be calculated repeatedly. It might be worth thinking about a way to skip the redundant calcuations.

Hallberg-NOAA commented:
Thank you for the revisions to this PR, @c2xu. The modifications to this PR have gone a long way toward addressing most of the big issues with the original commit.  However, there are a a number of specific comments (see above) requesting changes to the new MOM_harmonic_analysis module that have not been addressed.  Each of them are pretty localized and collectively they should only take a few minutes to address.  Please address each of those specific comments so that we can merge this PR into the MOM6 dev/gfdl branch.

c2xu commented:
> Thank you for the revisions to this PR, @c2xu. The modifications to this PR have gone a long way toward addressing most of the big issues with the original commit. However, there are a a number of specific comments (see above) requesting changes to the new MOM_harmonic_analysis module that have not been addressed. Each of them are pretty localized and collectively they should only take a few minutes to address. Please address each of those specific comments so that we can merge this PR into the MOM6 dev/gfdl branch.

Hi Bob, have you had a chance to look at the latest commit on May 6? (https://github.com/NOAA-GFDL/MOM6/pull/603/commits/1a0f375a1ba908a543edc3d36fd5d86fa56db7ca) It should have addressed all issues you raised. Please let me know if any further changes are required.

c2xu commented:
> Thank you for the revisions to this PR, @c2xu. The modifications to this PR have gone a long way toward addressing most of the big issues with the original commit. However, there are a a number of specific comments (see above) requesting changes to the new MOM_harmonic_analysis module that have not been addressed. Each of them are pretty localized and collectively they should only take a few minutes to address. Please address each of those specific comments so that we can merge this PR into the MOM6 dev/gfdl branch.

@Hallberg-NOAA, I just made some minor edits and I believe the code is ready for the merge now. Please let me know if you see any additional issues. Thanks.

Hallberg-NOAA reviewed (approved):
This PR has been modified and addresses (almost) all of the concerns that I had with the previous version.  There remains the challenge of how to get these diagnostics to work properly over an interval spanning a restart, but even without that enhancement to this new diagnostic capability, this is a valuable addition to the MOM6 code base, and I am approving this PR as currently written.

  We might want to add an issue noting the challenge of making the harmonic analysis work across restarts to remind ourselves to come back and revisit this at some point, or to provide some documentation to users of why this breaks the usual rules about exact reproducibility across restarts.

herrwang0 commented:
Apologies for the late response. I am still not quite sure about putting the calculation inside barotropic solver. I tend to think the 3D velocity at the end of baroclinic step is the "eventual" instantaneous velocity with that timestamp. It also gives you the flexibility to do harmonic analysis for non-barotropic fields. Another reason is the inconsistency of velocities between barotropic solver and baroclinic steps (at least for now).

Regardless, this is a nice addition to the model and the implementation seems largely correct. I agree with @Hallberg-NOAA that it should be added  as it is, and we can come back to address other issues in the future.

Hallberg-NOAA reviewed (changes requested):
We have hit one final (very minor snag) with the automated checking where the doxygen syntax describing range of the input data bounds (`is, ie, js, je`) in the `HA_type` is not quite right.  With all the variables on a single line, the trailing comment only applies to the last of these variables.  We could document each of these individually by putting them on separate lines, or we could document them together by revising 

> `integer :: is, ie, js, je !< Lower and upper bounds of input data`

into

> `!>@{ Lower and upper bounds of input data`
> `integer :: is, ie, js, je`
> `!>@}`

How we fix this (and get rid of the red-X for the TC testing) is up to you, @c2xu

c2xu commented:
> We have hit one final (very minor snag) with the automated checking where the doxygen syntax describing range of the input data bounds (`is, ie, js, je`) in the `HA_type` is not quite right. With all the variables on a single line, the trailing comment only applies to the last of these variables. We could document each of these individually by putting them on separate lines, or we could document them together by revising
> 
> > `integer :: is, ie, js, je !< Lower and upper bounds of input data`
> 
> into
> 
> > `!>@{ Lower and upper bounds of input data`
> > `integer :: is, ie, js, je`
> > `!>@}`
> 
> How we fix this (and get rid of the red-X for the TC testing) is up to you, @c2xu

This has been fixed.

c2xu commented:
> Apologies for the late response. I am still not quite sure about putting the calculation inside barotropic solver. I tend to think the 3D velocity at the end of baroclinic step is the "eventual" instantaneous velocity with that timestamp. It also gives you the flexibility to do harmonic analysis for non-barotropic fields. Another reason is the inconsistency of velocities between barotropic solver and baroclinic steps (at least for now).
> 
> Regardless, this is a nice addition to the model and the implementation seems largely correct. I agree with @Hallberg-NOAA that it should be added as it is, and we can come back to address other issues in the future.

This module provides the flexibility that `HA_register` and `HA_accum` can be called anywhere during the time stepping, and I agree there might be better place to call them, as long as `HA_init` is called first in `MOM_tidal_forcing`. One thing to note is that it currently works for 2D fields only. The compatibility with 3D fields should also be addressed in the future.

Hallberg-NOAA commented:
The latest update fixed the first set of doxygen errors, but now that we are past those, there are other errors that have been highlighted.  There need to be doxygen comments added to describe the `ha1` argument to `ha_write()` and the `ha1` and `nc` arguments to `ha_solver()`.  The doyxygen parse is also getting confused with the arguments to `ha_init()`.  All of the arguments there do have properly formatted comments, but I suspect that doxygen is getting confused by the line continuations between the declarations of `Time` and `time_ref` on lines 62-63 and between the dlclarations of `freq` and `phase0` on lines 67-68 of MOM_harmonic_analysis.F90

c2xu commented:
> The latest update fixed the first set of doxygen errors, but now that we are past those, there are other errors that have been highlighted. There need to be doxygen comments added to describe the `ha1` argument to `ha_write()` and the `ha1` and `nc` arguments to `ha_solver()`. The doyxygen parse is also getting confused with the arguments to `ha_init()`. All of the arguments there do have properly formatted comments, but I suspect that doxygen is getting confused by the line continuations between the declarations of `Time` and `time_ref` on lines 62-63 and between the dlclarations of `freq` and `phase0` on lines 67-68 of MOM_harmonic_analysis.F90

These comments are all fixed.

Hallberg-NOAA reviewed (changes requested):
I am sorry, but this commit is failing our coupled regression testing.  The problem is that the code in HA_init is looking for length of the run segment either from the contents of ocean_solo_nml or the runtime parameter `DAYMAX`.  These only exist and are used in ocean-only configurations.  In coupled configurations (either ice-ocean or fully coupled) the length of the run segments is determined by the coupler, as is the orchestration of the model termination.  As currently written, the ocean component is never informed of the length of a run segment when in coupled mode - it is just told by the coupler when to save its restart file (via a call to `ocean_model_restart()`) and terminate (via a call to `ocean_model_end()`).

  I fully understand why it makes sense for this new diagnostic to want to know whether it will work before it is triggered, and that it does not work yet with restarts, so it won't work with a sufficiently short run segment, but this will have to be revised to work with coupled configurations before we can accept it.

c2xu commented:
> I am sorry, but this commit is failing our coupled regression testing. The problem is that the code in HA_init is looking for length of the run segment either from the contents of ocean_solo_nml or the runtime parameter `DAYMAX`. These only exist and are used in ocean-only configurations. In coupled configurations (either ice-ocean or fully coupled) the length of the run segments is determined by the coupler, as is the orchestration of the model termination. As currently written, the ocean component is never informed of the length of a run segment when in coupled mode - it is just told by the coupler when to save its restart file (via a call to `ocean_model_restart()`) and terminate (via a call to `ocean_model_end()`).
> 
> I fully understand why it makes sense for this new diagnostic to want to know whether it will work before it is triggered, and that it does not work yet with restarts, so it won't work with a sufficiently short run segment, but this will have to be revised to work with coupled configurations before we can accept it.

So I guess the easiest fix is to add `HA_END_TIME` and let the users make sure the harmonic analysis is performed before the run segment ends.

c2xu commented:
> I am sorry, but this commit is failing our coupled regression testing. The problem is that the code in HA_init is looking for length of the run segment either from the contents of ocean_solo_nml or the runtime parameter `DAYMAX`. These only exist and are used in ocean-only configurations. In coupled configurations (either ice-ocean or fully coupled) the length of the run segments is determined by the coupler, as is the orchestration of the model termination. As currently written, the ocean component is never informed of the length of a run segment when in coupled mode - it is just told by the coupler when to save its restart file (via a call to `ocean_model_restart()`) and terminate (via a call to `ocean_model_end()`).
> 
> I fully understand why it makes sense for this new diagnostic to want to know whether it will work before it is triggered, and that it does not work yet with restarts, so it won't work with a sufficiently short run segment, but this will have to be revised to work with coupled configurations before we can accept it.

@Hallberg-NOAA I redefined the end time of harmonic analysis and fixed a related bug. Please let me know if this version works. Thanks!
